### ISSUE TITLE
Web workers do not work when runtime is edge

### ISSUE THREAD

**Issue by jauniusmentimeter** [vercel/next.js#67910]
### Link to the code that reproduces this issue

https://github.com/jauniusmentimeter/reproduction-app

### To Reproduce

1. Start the application with next dev command
2. Open the page http://localhost:3000
3. The error "document is not defined is returned from the server"
4. Open `app/page.tsx` and change runtime to `nodejs` instead of `edge`. The error is not returned anymore.

### Current vs. Expected behavior

Web workers do not work when runtime is set to edge. They should work in the same way as on nodejs runtime, since they are running on the client.

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.5.0: Wed May  1 20:14:38 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6020
  Available memory (MB): 16384
  Available CPU cores: 10
Binaries:
  Node: 20.11.0
  npm: 9.5.1
  Yarn: 1.22.21
  pnpm: 9.5.0
Relevant Packages:
  next: 15.0.0-canary.72 // Latest available version is detected (15.0.0-canary.72).
  eslint-config-next: N/A
  react: 19.0.0-rc.0
  react-dom: 19.0.0-rc.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Runtime, Webpack

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local), next start (local), Vercel (Deployed)

### Additional context

For some reason document is trying to be accessed on server environment when web worker code exist:
<img width="1021" alt="image" src="https://github.com/user-attachments/assets/1f2f4d0d-b642-4308-a6ff-47c05894370f">
here is the place in `.next/server/edge-runtime-webpack.js` that is failing:
<img width="704" alt="image" src="https://github.com/user-attachments/assets/8312efc0-6373-48e6-af86-96d44c7b7921">

**Comment by JesseKoldewijn** [comment]
Odd indeed that this behaviour comes up like this

**Comment by icyJoseph** [comment]
Does it work if you do:

```js
   const worker = new Worker(new URL('./worker.js', new URL(import.meta.url)));

```

**Comment by jauniusmentimeter** [comment]
> Does it work if you do:
> 
> ```js
>    const worker = new Worker(new URL('./worker.js', new URL(import.meta.url)));
> ```
> 
> ### Edit
> Err, that shadows the error I think, never mind...

yeah that will not work. I'd love to have some workaround for this issue though if possible 👀

**Comment by icyJoseph** [comment]
I guess you could output your worker to another folder, but 🤔 perhaps not ideal... Like they do here, https://medium.com/@ngrato/harnessing-the-power-of-web-workers-with-next-js-350901a99a10

**Comment by nickbabcock** [comment]
Some notes:
- The error only happens in development. A production build is fine.
- Switching to turbopack for development also fixes the issue  (eg: `next dev --turbo`)
- Building the worker separately and storing in `public/` does also work, but the thought of non-digested static js asset may be a tough sell. Seems possible to workaround by creating a webpack asset module special case for our worker, but might be more effort than it's worth.
- Seems like the issue has always been present, from Next.js 13.4 to 15.0.0-canary.111

**Comment by kyb3r** [comment]
This issue is still present in the latest next version. Any workarounds would be appreciated, since we're gonna be stuck on next 14, since next 15 has huge breaking changes.

> Building the worker separately and storing in public/ does also work, but the thought of non-digested static js asset may be a tough sell. Seems possible to workaround by creating a webpack asset module special case for our worker, but might be more effort than it's worth.

Are there any examples of how to do this?

**Comment by kyb3r** [comment]
> * The error only happens in development. A production build is fine.

Production build just failed for me with the same error. I have to switch to using the nodejs runtime although that is not preferred for me.
